File: scripts/filter/Raster.py

import pyproj
import math
from utils import int2deg, deg2int

scale = 10

class Raster:
	""" Class for representing raster over a map"""	
	dimension = 20
	minlon = 0
	minlat = 0
	raster = [[]]
	lonparts = 0
	latparts = 0

	def __init__(self,amap):
		""" Creates new raster from given map"""
		self.minlon = 10**10
		self.minlat = 10**10
		maxlon = 0
		maxlat = 0
		for node in amap.nodes:
			self.minlon = min(self.minlon,node.lon)
			self.minlat = min(self.minlat,node.lat)
			maxlon = max(maxlon,node.lon)
			maxlat = max(maxlat,node.lat)
		
		dlon = maxlon-self.minlon
		dlat = maxlat-self.minlat

		londist = dlon/scale
		latdist = dlat/scale
		
		print "Lon:",self.minlon," -- ",maxlon
		print "Lat:",self.minlat," -- ",maxlat

		self.lonparts = int(math.ceil(londist/self.dimension))+10
		self.latparts = int(math.ceil(latdist/self.dimension))+10

		print "Creating raster",self.lonparts,"x",self.latparts

		self.steplon = int(math.ceil(1.0*dlon/self.lonparts))
		self.steplat = int(math.ceil(1.0*dlat/self.latparts))
		self.raster = [[[] for j in range(self.latparts)] for i in range(self.lonparts)]
		for node in amap.nodes:
			self.raster[(node.lon-self.minlon)/self.steplon][(node.lat-self.minlat)/self.steplat].append(node.id)

	def getBox(self,lon,lat):
		""" Get coordinates of a raster box in which given coordinates lies"""
		return ((lon-self.minlon)/self.steplon,(lat-self.minlat)/self.steplat)

File: compiled/raster.c
#include <ucw/lib.h>
#include <stdlib.h>
#include <ucw/gary.h>

#include "types.h"
#include "raster.h"
#include "hashes.c"
#include "utils.h"


struct raster_t makeRaster(struct map_t map){
	int dimension = 20;

	struct raster_t raster;
	
	int64_t maxlon;
	int64_t maxlat;

	raster.minlon=map.nodes[0]->lon;
	raster.minlat=map.nodes[0]->lat;
	maxlon=map.nodes[0]->lon;
	maxlat=map.nodes[0]->lat;
	
	for (int i=0;i<map.n_nodes;i++){
		raster.minlon = MIN(map.nodes[i]->lon,raster.minlon);	
		maxlon = MAX(map.nodes[i]->lon,maxlon);	
		raster.minlat = MIN(map.nodes[i]->lat,raster.minlat);
		maxlat = MAX(map.nodes[i]->lat,maxlat);
	}

	int64_t dlon = maxlon-raster.minlon;
	int64_t dlat = maxlat-raster.minlat;

	printf("Lon: %d -- %d, Lat: %d -- %d\n",raster.minlon,maxlon,raster.minlat,maxlat);

	double londist;
	double latdist;
	
	londist = int2deg(maxlon-raster.minlon);
	latdist = int2deg(maxlat-raster.minlat);
	
	raster.lonparts = (int)(londist/dimension)+10;
	raster.latparts = (int)(latdist/dimension)+10;
	
	printf("Creating raster %d x %d\n",raster.lonparts,raster.latparts);

	raster.steplon = (dlon+raster.lonparts-1)/raster.lonparts;
	raster.steplat = (dlat+raster.latparts-1)/raster.latparts;

	GARY_INIT(raster.raster,raster.lonparts);
	for (int i=0;i<raster.lonparts;i++){
		GARY_INIT(raster.raster[i],raster.latparts);
		for (int j=0;j<raster.latparts;j++){
			GARY_INIT(raster.raster[i][j],0);	
		}
	}

	for (int i=0;i<map.n_nodes;i++){
		Premap__Node * node;
		node = map.nodes[i];
		int * ptr;
		ptr = GARY_PUSH(raster.raster[(node->lon-raster.minlon)/raster.steplon][(node->lat-raster.minlat)/raster.steplat]);
		* ptr = nodesIdx_find(node->id)->idx;
	}

	return raster;
}

int * getRasterBox(struct raster_t raster,int64_t lon, int64_t lat){
	int * box;
	box = malloc(sizeof(int)*2);
	box[0]=(lon-raster.minlon)/raster.steplon;
	box[1]=(lat-raster.minlat)/raster.steplat;	
	return box;
}
